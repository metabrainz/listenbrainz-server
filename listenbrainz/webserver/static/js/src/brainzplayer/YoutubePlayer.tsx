import * as React from "react";
import YouTube, { Options } from "react-youtube";
import {
  isEqual as _isEqual,
  get as _get,
  isNil as _isNil,
  isString as _isString,
  isFunction as _isFunction,
} from "lodash";
import { DataSourceType, DataSourceProps } from "./BrainzPlayer";
import {
  getArtistName,
  getTrackName,
  searchForYoutubeTrack,
} from "../utils/utils";

type YoutubePlayerState = {
  currentListen?: Listen;
};

type YoutubePlayerProps = DataSourceProps & {
  youtubeUser?: YoutubeUser;
  refreshYoutubeToken: () => Promise<string>;
};

// For some reason Youtube types do not document getVideoData,
// which we need to determine if there was no search results
type ExtendedYoutubePlayer = {
  getVideoData?: () => { video_id?: string; author: string; title: string };
} & YT.Player;

export default class YoutubePlayer
  extends React.Component<YoutubePlayerProps, YoutubePlayerState>
  implements DataSourceType {
  static getVideoIDFromListen(listen: Listen | JSPFTrack): string | undefined {
    // Checks if there is a youtube ID in the listen
    // or if the origin_url field contains youtube.com
    const videoURL =
      _get(listen, "track_metadata.additional_info.youtube_id") ??
      _get(listen, "track_metadata.additional_info.origin_url");
    if (_isString(videoURL) && videoURL.length) {
      /** Credit for this regular expression goes to Soufiane Sakhi:
       * https://stackoverflow.com/a/61033353/4904467
       */
      const youtubeURLRegexp = /(?:https?:\/\/)?(?:www\.|m\.)?youtu(?:\.be\/|be.com\/\S*(?:watch|embed)(?:(?:(?=\/[^&\s?]+(?!\S))\/)|(?:\S*v=|v\/)))([^&\s?]+)/g;
      const match = youtubeURLRegexp.exec(videoURL);
      return match?.[1];
    }

    return undefined;
  }

  static isListenFromThisService(listen: Listen | JSPFTrack): boolean {
    const videoId = YoutubePlayer.getVideoIDFromListen(listen);
    return Boolean(videoId);
  }

  /**
   * Youtube thumbnail URLs can be composed with <video_id>/<resolution><image>.jpg
   * where resolution is one of [hq , md, sd] and image is either 'default' (= 0)
   * or a number between 0 -> 3 (there are 4 thumbnails)
   */
  static getThumbnailsFromVideoid(videoId?: string) {
    let images: MediaImage[] = [];
    if (videoId) {
      images = [
        {
          src: `http://img.youtube.com/vi/${videoId}/sddefault.jpg`,
          sizes: "640x480",
          type: "image/jpg",
        },
        {
          src: `http://img.youtube.com/vi/${videoId}/hqdefault.jpg`,
          sizes: "480x360",
          type: "image/jpg",
        },
        {
          src: `http://img.youtube.com/vi/${videoId}/mqdefault.jpg`,
          sizes: "320x180",
          type: "image/jpg",
        },
      ];
    }
    return images;
  }

  static getYoutubeURLFromListen(
    listen: Listen | JSPFTrack
  ): string | undefined {
    // Checks if there is a youtube ID in the listen
    const youtubeId = _get(listen, "track_metadata.additional_info.youtube_id");
    if (youtubeId) {
      return `https://www.youtube.com/watch?v=${youtubeId}`;
    }

    // or if the origin URL contains youtube.com
    const originURL = _get(listen, "track_metadata.additional_info.origin_url");
    if (_isString(originURL) && originURL.length) {
      try {
        const parsedURL = new URL(originURL);
        const { hostname, searchParams } = parsedURL;
        if (/youtube\.com/.test(hostname)) {
          return originURL;
        }
      } catch {
        return undefined;
      }
    }
    return undefined;
  }

  public name = "youtube";
  public domainName = "youtube.com";
  youtubePlayer?: ExtendedYoutubePlayer;
  checkVideoLoadedTimerId?: NodeJS.Timeout;

  componentDidUpdate(prevProps: DataSourceProps) {
    const { show } = this.props;
    if (prevProps.show && !show && this.youtubePlayer) {
      this.youtubePlayer.stopVideo();
      // Clear playlist
      this.youtubePlayer.cueVideoById("");
    }
  }

  onReady = (event: YT.PlayerEvent): void => {
    this.youtubePlayer = event.target;
  };

  updateVideoInfo = (): void => {
    let title;
    let images: MediaImage[] = [];
    const { onTrackInfoChange, onDurationChange } = this.props;
    const videoData =
      this.youtubePlayer?.getVideoData && this.youtubePlayer.getVideoData();
    let videoId: string = "";
    if (videoData) {
      title = videoData.title;
      videoId = videoData.video_id as string;
      images = YoutubePlayer.getThumbnailsFromVideoid(videoId);
    } else {
      // Fallback to track name from the listen we are playing
      const { currentListen } = this.state;
      title = getTrackName(currentListen);
    }
    onTrackInfoChange(
      title,
      `https://www.youtube.com/watch?v=${videoId}`,
      undefined,
      undefined,
      images
    );
    const duration = this.youtubePlayer?.getDuration();
    if (duration) {
      onDurationChange(duration * 1000);
    }
  };

  handlePlayerStateChanged = (event: YT.OnStateChangeEvent) => {
    const { data: state, target: player } = event;
    const {
      onPlayerPausedChange,
      onDurationChange,
      onProgressChange,
      onTrackInfoChange,
      show,
    } = this.props;
    if (!show) {
      return;
    }
    if (state === YouTube.PlayerState.ENDED) {
      const { onTrackEnd } = this.props;
      onTrackEnd();
      return;
    }
    // New track being played
    if (state === YouTube.PlayerState.UNSTARTED) {
      const title = _get(player, "playerInfo.videoData.title", "");
      const videoId = _get(player, "playerInfo.videoData.video_id", "");
      // The player info is sometimes missing a title initially.
      // We fallback to getting it with getVideoData method once the information is loaded in the player
      if (!title) {
        setTimeout(this.updateVideoInfo.bind(this), 2000);
      } else {
        const images: MediaImage[] = YoutubePlayer.getThumbnailsFromVideoid(
          videoId
        );
        onTrackInfoChange(title, videoId, undefined, undefined, images);
      }
      player.playVideo();
    }
    if (
      state === YouTube.PlayerState.UNSTARTED ||
      state === YouTube.PlayerState.BUFFERING
    ) {
      onPlayerPausedChange(false);
    }
    if (state === YouTube.PlayerState.PAUSED) {
      onPlayerPausedChange(true);
    }
    if (state === YouTube.PlayerState.PLAYING) {
      onPlayerPausedChange(false);
    }
    onProgressChange(player.getCurrentTime() * 1000);
    const duration = _isFunction(player.getDuration) && player.getDuration();
    if (duration) {
      onDurationChange(duration * 1000);
    }
  };

  handleAccountError = (): void => {
    const errorMessage = (
      <p>
        In order to play music with Youtube, you will need a Youtube / Google
        account linked to your ListenBrainz account.
        <br />
        Please try to{" "}
        <a href="/profile/music-services/details/" target="_blank">
          link for &quot;playing music&quot; feature
        </a>{" "}
        and refresh this page
      </p>
    );
    const { onTrackNotFound, handleWarning } = this.props;
    handleWarning(errorMessage);
    onTrackNotFound();
  };

  searchAndPlayTrack = async (listen: Listen | JSPFTrack) => {
    const trackName = getTrackName(listen);
    const artistName = getArtistName(listen);
    // Using the releaseName has paradoxically given worst search results,
    // so we're only using it when track name isn't provided (for example for an album search)
    const releaseName = trackName
      ? ""
      : _get(listen, "track_metadata.release_name");

    const {
      handleWarning,
      onTrackNotFound,
      youtubeUser,
      refreshYoutubeToken,
    } = this.props;

    if (!this.youtubePlayer) {
      onTrackNotFound();
      return;
    }
    // If the user is not authed for Youtube API, show a helpful error message with link to connect
    if (!youtubeUser) {
      this.handleAccountError();
      return;
    }
    if (!trackName && !artistName && !releaseName) {
      handleWarning("Not enough info to search on Youtube");
      onTrackNotFound();
      return;
    }

    try {
      const { api_key } = youtubeUser;
      const videoIds = await searchForYoutubeTrack(
        api_key,
        trackName,
        artistName,
        releaseName,
        refreshYoutubeToken,
        this.handleAccountError
      );
      if (videoIds?.length) {
        this.playTrackById(videoIds[0]);
      } else {
        onTrackNotFound();
      }
    } catch (error) {
      handleWarning(error?.message ?? error.toString(), "Youtube player error");
      onTrackNotFound();
    }
  };

  playTrackById = (videoId: string): void => {
    if (!videoId || !this.youtubePlayer) {
      return;
    }
    if (videoId.startsWith("http")) {
      this.youtubePlayer.loadVideoByUrl(videoId);
    } else {
      this.youtubePlayer.loadVideoById(videoId);
    }
  };

  canSearchAndPlayTracks = (): boolean => {
    const { youtubeUser } = this.props;
    // check if the user is authed to search with the Youtube API
    return Boolean(youtubeUser) && Boolean(youtubeUser?.api_key);
  };

  datasourceRecordsListens = (): boolean => {
    return false;
  };

  playListen = (listen: Listen | JSPFTrack) => {
    const { show } = this.props;
    if (!show) {
      return;
    }
    let youtubeId = _get(listen, "track_metadata.additional_info.youtube_id");
    const originURL = _get(listen, "track_metadata.additional_info.origin_url");
    if (!youtubeId && _isString(originURL) && originURL.length) {
      try {
        const parsedURL = new URL(originURL);
        const { hostname, searchParams } = parsedURL;
        if (/youtube\.com/.test(hostname)) {
          youtubeId = searchParams.get("v");
        }
      } catch {
        // URL is not valid, do nothing
      }
    }
    if (youtubeId) {
      this.playTrackById(youtubeId);
    } else {
      this.searchAndPlayTrack(listen);
    }
  };

  togglePlay = (): void => {
    if (!this.youtubePlayer) {
      return;
    }
    const { playerPaused, onPlayerPausedChange } = this.props;
    if (playerPaused) {
      this.youtubePlayer.playVideo();
      onPlayerPausedChange(false);
    } else {
      this.youtubePlayer.pauseVideo();
      onPlayerPausedChange(true);
    }
  };

  seekToPositionMs = (msTimecode: number) => {
    if (!this.youtubePlayer) {
      return;
    }
    this.youtubePlayer.seekTo(msTimecode / 1000, true);
    this.youtubePlayer.playVideo();
  };

  onError = (event: YT.OnErrorEvent): void => {
    const { data: errorNumber } = event;
    const { handleError, onTrackNotFound } = this.props;
    if (this.checkVideoLoadedTimerId) {
      clearTimeout(this.checkVideoLoadedTimerId);
    }
    let message = "Something went wrong";
    switch (errorNumber) {
      case 101:
      case 150:
        message =
          "The owner of the requested video does not allow it to be played in embedded players.";
        break;
      case 5:
        message = "The requested content cannot be played in an HTML5 player.";
        break;
      case 2:
        message = "The request contained an invalid parameter value.";
        break;
      case 100:
        message = "The video requested was not found.";
        break;
      default:
        break;
    }
    handleError(message, "Youtube player error");
    onTrackNotFound();
  };

  render() {
    const { show } = this.props;
    const options: Options = {
      playerVars: {
        autoplay: 1,
        controls: 0,
        showinfo: 0,
        fs: 0,
        iv_load_policy: 3,
        modestbranding: 1,
        rel: 0,
        origin: window.location.origin,
      },
      width: "100%",
      height: "100%",
    };
    return (
      <div className={`youtube ${!show ? "hidden" : ""}`}>
        <YouTube
          opts={options}
          onError={this.onError}
          onStateChange={this.handlePlayerStateChanged}
          onReady={this.onReady}
        />
      </div>
    );
  }
}
