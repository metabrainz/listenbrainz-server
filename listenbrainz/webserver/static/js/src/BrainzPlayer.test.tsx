import * as React from "react";
import { shallow, mount } from "enzyme";
import BrainzPlayer, { DataSourceType } from "./BrainzPlayer";
import SoundcloudPlayer from "./SoundcloudPlayer";
import YoutubePlayer from "./YoutubePlayer";
import SpotifyPlayer from "./SpotifyPlayer";
import APIService from "./APIService";

const props = {
  spotifyUser: {
    access_token: "heyo",
    permission: "read" as SpotifyPermission,
  },
  direction: "up" as BrainzPlayDirection,
  onCurrentListenChange: (listen: Listen | JSPFTrack) => {},
  listens: [],
  newAlert: (
    alertType: AlertType,
    title: string,
    message: string | JSX.Element
  ) => {},
  apiService: new APIService("base-uri"),
};

// Give yourself a two minute break and go listen to this gem
// https://musicbrainz.org/recording/7fcaf5b3-e682-4ce6-be61-d3bce775a43f
const listen: Listen = {
  listened_at: 0,
  track_metadata: {
    artist_name: "Moondog",
    track_name: "Bird's Lament",
  },
};

describe("BrainzPlayer", () => {
  it("renders correctly", () => {
    const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("creates Youtube and SoundCloud datasources by default", () => {
    const mockProps = {
      ...props,
      spotifyUser: {},
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    expect(instance.dataSources).toHaveLength(2);
    expect(instance.dataSources[0].current).toBeInstanceOf(YoutubePlayer);
    expect(instance.dataSources[1].current).toBeInstanceOf(SoundcloudPlayer);
  });

  it("creates a Spotify datasource when passed a spotify user with right permissions", () => {
    const mockProps = {
      ...props,
      spotifyUser: {
        access_token: "haveyouseenthefnords",
        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
      },
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    expect(instance.dataSources[0].current).toBeInstanceOf(SpotifyPlayer);
  });

  it("removes a datasource when calling invalidateDataSource", () => {
    const mockProps = {
      ...props,
      spotifyUser: {
        access_token: "haveyouseenthefnords",
        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
      },
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    instance.handleWarning = jest.fn();

    const datasourcesBefore = instance.dataSources.length;
    instance.invalidateDataSource(
      instance.dataSources[0].current as SpotifyPlayer,
      "Test message"
    );
    expect(instance.handleWarning).toHaveBeenCalledWith(
      "Test message",
      "Cannot play from this source"
    );
    expect(instance.dataSources).toHaveLength(datasourcesBefore - 1);
    instance.dataSources.forEach((dataSource) => {
      expect(dataSource.current).not.toBeInstanceOf(SpotifyPlayer);
    });
  });

  it("selects Youtube as source when listen has a youtube URL", () => {
    const mockProps = {
      ...props,
      spotifyUser: {
        access_token: "haveyouseenthefnords",
        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
      },
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    expect(instance.dataSources[1].current).toBeInstanceOf(YoutubePlayer);
    const youtubeListen: Listen = {
      listened_at: 0,
      track_metadata: {
        artist_name: "Moondog",
        track_name: "Bird's Lament",
        additional_info: {
          origin_url: "https://www.youtube.com/watch?v=RW8SBwGNcF8",
        },
      },
    };
    // if origin_url is a youtube link, it should play it with YoutubePlayer instead of Spotify
    instance.playListen(youtubeListen);
    expect(instance.state.currentDataSourceIndex).toEqual(1);
  });

  it("selects Spotify as source when listen has listening_from = spotify", () => {
    const mockProps = {
      ...props,
      spotifyUser: {
        access_token: "haveyouseenthefnords",
        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
      },
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    const spotifyListen: Listen = {
      listened_at: 0,
      track_metadata: {
        artist_name: "Moondog",
        track_name: "Bird's Lament",
        additional_info: {
          listening_from: "spotify",
        },
      },
    };
    // Try to play on youtube directly (index 1), should use spotify instead (index 0)
    instance.playListen(spotifyListen, 1);
    expect(instance.state.currentDataSourceIndex).toEqual(0);
  });

  it("selects Spotify as source when listen has a spotify_id", () => {
    const mockProps = {
      ...props,
      spotifyUser: {
        access_token: "haveyouseenthefnords",
        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
      },
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    const spotifyListen: Listen = {
      listened_at: 0,
      track_metadata: {
        artist_name: "Moondog",
        track_name: "Bird's Lament",
        additional_info: {
          spotify_id: "doesn't matter in this test as long as it's here",
        },
      },
    };
    // Try to play on youtube directly (index 1), should use spotify instead (index 0)
    instance.playListen(spotifyListen, 1);
    expect(instance.state.currentDataSourceIndex).toEqual(0);
  });

  it("selects Soundcloud as source when listen has a soundcloud URL", () => {
    const mockProps = {
      ...props,
      spotifyUser: {
        access_token: "haveyouseenthefnords",
        permission: "streaming user-read-email user-read-private" as SpotifyPermission,
      },
    };
    const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
    const instance = wrapper.instance();
    expect(instance.dataSources[2].current).toBeInstanceOf(SoundcloudPlayer);
    const soundcloudListen: Listen = {
      listened_at: 42,
      track_metadata: {
        additional_info: {
          origin_url: "https://soundcloud.com/wankelmut/wankelmut-here-to-stay",
        },
        artist_name: "Wankelmut",
        release_name: "",
        track_name: "Rock'n'Roll Is Here To Stay",
      },
    };
    // if origin_url is a soundcloud link, it should play it with SoundcloudPlayer instead of Spotify
    instance.playListen(soundcloudListen);
    expect(instance.state.currentDataSourceIndex).toEqual(2);
  });

  describe("isCurrentListen", () => {
    it("returns true if currentListen and passed listen is same", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();

      wrapper.setProps({ currentListen: listen });

      expect(instance.isCurrentListen(listen)).toBe(true);
    });

    it("returns false if currentListen is not set", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();

      wrapper.setProps({ currentListen: undefined });

      expect(instance.isCurrentListen({} as Listen)).toBeFalsy();
    });
  });

  describe("getCurrentTrackName", () => {
    it("returns the track name when it exists on a listen", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();

      wrapper.setProps({ currentListen: listen });

      expect(instance.getCurrentTrackName()).toEqual("Bird's Lament");
    });

    it("returns an empty string if currentListen is not set", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();

      wrapper.setProps({ currentListen: undefined });

      expect(instance.getCurrentTrackName()).toEqual("");
    });
  });

  describe("getCurrentTrackArtists", () => {
    it("returns the track artists string when it exists on a listen", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();

      wrapper.setProps({ currentListen: listen });

      expect(instance.getCurrentTrackArtists()).toEqual("Moondog");
    });

    it("returns an empty string if currentListen is not set", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();

      wrapper.setProps({ currentListen: undefined });

      expect(instance.getCurrentTrackArtists()).toEqual("");
    });
  });
  describe("seekToPositionMs", () => {
    it("invalidates the datasource if it doesn't exist", () => {
      const mockProps = {
        ...props,
        spotifyUser: {},
      };
      const wrapper = mount<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
      const instance = wrapper.instance();

      const fakeDatasource = {
        current: false,
      };
      instance.dataSources.push(fakeDatasource as any);

      // Setting non-existing datasource index
      const datasourcesBefore = instance.dataSources.length;
      wrapper.setState({ currentDataSourceIndex: datasourcesBefore - 1 });
      expect(
        instance.dataSources[instance.state.currentDataSourceIndex].current
      ).toEqual(false);

      instance.invalidateDataSource = jest.fn(() =>
        instance.dataSources.splice(instance.state.currentDataSourceIndex, 1)
      );
      instance.seekToPositionMs(1000);

      expect(instance.invalidateDataSource).toHaveBeenCalledTimes(1);
      expect(instance.invalidateDataSource).toHaveBeenCalledWith();
      expect(instance.dataSources).toHaveLength(datasourcesBefore - 1);
      // Check that it removed the right datasource
      instance.dataSources.forEach((dataSource) => {
        expect(dataSource.current).not.toEqual(false);
      });
    });

    it("calls seekToPositionMs on the datasource", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();
      instance.invalidateDataSource = jest.fn();
      const fakeDatasource = {
        current: {
          seekToPositionMs: jest.fn(),
        },
      };
      instance.dataSources = [fakeDatasource as any];
      instance.seekToPositionMs(1000);
      expect(instance.invalidateDataSource).not.toHaveBeenCalled();
      expect(fakeDatasource.current.seekToPositionMs).toHaveBeenCalledTimes(1);
      expect(fakeDatasource.current.seekToPositionMs).toHaveBeenCalledWith(
        1000
      );
    });
  });
  describe("toggleDirection", () => {
    it("sets direction to 'down' if not set to a recognised value", () => {
      const mockProps = {
        ...props,
        direction: "" as BrainzPlayDirection,
      };
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
      const instance = wrapper.instance();
      expect(instance.state.direction).toEqual("down");
      instance.setState({ direction: "fnord" as BrainzPlayDirection });
      expect(instance.state.direction).toEqual("fnord");
      instance.toggleDirection();
      expect(instance.state.direction).toEqual("down");
    });

    it("alternates direction between 'up' and 'down'", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();
      expect(instance.state.direction).toEqual("up");
      instance.toggleDirection();
      expect(instance.state.direction).toEqual("down");
      instance.toggleDirection();
      expect(instance.state.direction).toEqual("up");
    });
  });

  describe("failedToFindTrack", () => {
    it("calls playNextTrack if currentListen is not set", () => {
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...props} />);
      const instance = wrapper.instance();
      instance.playNextTrack = jest.fn();
      instance.failedToFindTrack();
      expect(instance.playNextTrack).toHaveBeenCalledTimes(1);
    });

    it("tries playing the current listen with the next datasource", () => {
      const mockProps = {
        ...props,
        currentListen: listen,
      };
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
      const instance = wrapper.instance();
      instance.playListen = jest.fn();
      instance.failedToFindTrack();
      expect(instance.playListen).toHaveBeenCalledTimes(1);
      expect(instance.playListen).toHaveBeenCalledWith(listen, 1);
    });

    it("shows a warning if out of datasources and plays next track", () => {
      const mockProps = {
        ...props,
        currentListen: listen,
      };
      const wrapper = shallow<BrainzPlayer>(<BrainzPlayer {...mockProps} />);
      const instance = wrapper.instance();
      instance.playNextTrack = jest.fn();
      instance.handleWarning = jest.fn();
      instance.setState({
        currentDataSourceIndex: instance.dataSources.length - 1,
      });
      instance.failedToFindTrack();
      expect(instance.playNextTrack).toHaveBeenCalledTimes(1);
      expect(instance.handleWarning).toHaveBeenCalledTimes(1);
      expect(instance.handleWarning).toHaveBeenCalledWith(
        "We couldn't find a matching song on any music service we tried",
        "Oh no !"
      );
    });
  });
});
