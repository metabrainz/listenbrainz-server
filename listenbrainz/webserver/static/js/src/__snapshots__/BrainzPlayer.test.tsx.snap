// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrainzPlayer renders correctly 1`] = `"<div><div id=\\"music-player\\" aria-label=\\"Playback control\\"><div class=\\"content\\"><div class=\\"youtube hidden\\"><div class=\\"\\"><div></div></div></div><div class=\\"soundcloud hidden\\"><iframe id=\\"soundcloud-iframe\\" title=\\"Soundcloud player\\" width=\\"100%\\" height=\\"420px\\" scrolling=\\"no\\" frameborder=\\"no\\" allow=\\"autoplay\\" src=\\"https://w.soundcloud.com/player/?url=\\"></iframe></div><div class=\\"no-album-art\\"></div></div><div class=\\"info showControls\\"><div class=\\"currently-playing\\"><div class=\\"progress\\" aria-label=\\"Audio Progress Control\\" role=\\"progressbar\\" aria-valuemin=\\"0\\" aria-valuemax=\\"100\\" aria-valuenow=\\"NaN\\" tabindex=\\"0\\"><div class=\\"progress-bar\\"></div></div></div><div class=\\"controls\\"><div class=\\"left btn btn-xs\\" title=\\"Always show controls\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"eye-slash\\" class=\\"svg-inline--fa fa-eye-slash fa-w-20 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 640 512\\"><path fill=\\"currentColor\\" d=\\"M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z\\"></path></svg></div><div class=\\"previous btn btn-xs\\" title=\\"Previous\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"fast-backward\\" class=\\"svg-inline--fa fa-fast-backward fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M0 436V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12v151.9L235.5 71.4C256.1 54.3 288 68.6 288 96v131.9L459.5 71.4C480.1 54.3 512 68.6 512 96v320c0 27.4-31.9 41.7-52.5 24.6L288 285.3V416c0 27.4-31.9 41.7-52.5 24.6L64 285.3V436c0 6.6-5.4 12-12 12H12c-6.6 0-12-5.4-12-12z\\"></path></svg></div><div class=\\"play btn\\" title=\\"Play\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"play-circle\\" class=\\"svg-inline--fa fa-play-circle fa-w-16 fa-2x \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm115.7 272l-176 101c-15.8 8.8-35.7-2.5-35.7-21V152c0-18.4 19.8-29.8 35.7-21l176 107c16.4 9.2 16.4 32.9 0 42z\\"></path></svg></div><div class=\\"next btn btn-xs\\" title=\\"Next\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"fast-forward\\" class=\\"svg-inline--fa fa-fast-forward fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M512 76v360c0 6.6-5.4 12-12 12h-40c-6.6 0-12-5.4-12-12V284.1L276.5 440.6c-20.6 17.2-52.5 2.8-52.5-24.6V284.1L52.5 440.6C31.9 457.8 0 443.4 0 416V96c0-27.4 31.9-41.7 52.5-24.6L224 226.8V96c0-27.4 31.9-41.7 52.5-24.6L448 226.8V76c0-6.6 5.4-12 12-12h40c6.6 0 12 5.4 12 12z\\"></path></svg></div><div class=\\"right btn btn-xs\\" title=\\"Play down\\" role=\\"button\\" tabindex=\\"0\\"><svg aria-hidden=\\"true\\" focusable=\\"false\\" data-prefix=\\"fas\\" data-icon=\\"sort-amount-up\\" class=\\"svg-inline--fa fa-sort-amount-up fa-w-16 \\" role=\\"img\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 512 512\\"><path fill=\\"currentColor\\" d=\\"M304 416h-64a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h64a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM16 160h48v304a16 16 0 0 0 16 16h32a16 16 0 0 0 16-16V160h48c14.21 0 21.38-17.24 11.31-27.31l-80-96a16 16 0 0 0-22.62 0l-80 96C-5.35 142.74 1.77 160 16 160zm416 0H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h192a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zm-64 128H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h128a16 16 0 0 0 16-16v-32a16 16 0 0 0-16-16zM496 32H240a16 16 0 0 0-16 16v32a16 16 0 0 0 16 16h256a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z\\"></path></svg></div></div></div></div></div>"`;
